File: python/function.py
import json
import logging
import os
import requests
from newrelic import agent

# In a python Lambda, the runtime loads the handler code as a module; so code in the top level
# of the module occurs once, during cold start.
print("Lambda Handler starting up")

# configure log level
logging.basicConfig(level=logging.INFO)

@agent.profile_trace()
def get_user():
    print("getting DB username ")
    return "Bill"

@agent.function_trace()
def get_pass():
    print("getting DB password")
    return "123456"

@agent.lambda_handler()
async def lambda_handler(event, context):
    # At this point, we're handling an invocation. Cold start is over; this code runs for each invocation.

    # Get the transaction object for this function
    transaction = agent.current_transaction()

    # Accept distributed tracing headers
    headers = event.get("headers", {})
    if headers:
        try:
            transaction.accept_distributed_trace_headers(headers)
        except Exception as e:
            logging.error("An exception occurred accepting dt headers: %s", e, exc_info=True)
    else:
        try:
            # Generate distributed tracing headers if headers are not provided
            headers_list = []
            transaction.insert_distributed_trace_headers(headers_list)
            headers = {k: v for k, v in headers_list}
        except Exception as e:
            logging.error("An exception occurred inserting dt headers: %s", e, exc_info=True)

    # Print out the distributed tracing headers
    print("Distributed tracing headers:")
    logging.info("The proprietary `newrelic` header can be decoded with: `pbpaste | base64 -d | jq .`")
    try:
        for key, value in headers.items():
            print(f"{key}: {value}")
    except Exception as e:
        logging.error("An exception occurred printing dt headers: %s", e, exc_info=True)

    # Make an external HTTP request and inject distributed trace headers
    response = requests.get("https://example.com", headers=headers)

    # additional function processes
    files = os.listdir("/opt/python/")
    with open("/opt/python/package-lock.json", "r") as f:
        package_lock = f.read()
    username = get_user()
    password = get_pass()

    # As normal, anything you write to stdout ends up in CloudWatch
    logging.info("username: ", username)
    logging.info("password: ", password)
    logging.info("ENVIRONMENT VARIABLES\n" + json.dumps(dict(os.environ), indent=2))
    logging.info("EVENT\n" + json.dumps(event, indent=2))
    logging.warning("This is a warning log.")
    logging.error("This is an error log!")
    logging.info(files)
    logging.info("***package-lock.json***\n", package_lock)
    logging.info("logStream: ", context.log_stream_name)

    return response.status_code

File: nodejs/function.js
const axios = require("axios")
const newrelic = require("newrelic")
const fs = require("fs")

// In a Node Lambda, the runtime loads the handler code as a module; so code in the top level
// of the module occurs once, during cold start.
console.log("Lambda Handler starting up")

function get_user() {
  console.log("getting DB username ")
  return "Bill"
}

function get_pass() {
  console.log("getting DB password")
  return "123456"
}

module.exports.lambda_handler = async function (event, context) {
  // Call newrelic.getTransaction to retrieve a handle on the current transaction.
  const transaction = newrelic.getTransaction()

  // Accept distributed tracing headers
  let headers = event.headers || {}
  if (!headers.hasOwnProperty("newrelic")) {
    // Generate distributed tracing headers if headers are not provided
    newrelic.startBackgroundTransaction(
      "background task",
      function executeTransaction() {
        const backgroundT = newrelic.getTransaction()
        // Generate the headers
        backgroundT.insertDistributedTraceHeaders(headers)
      }
    )
  } else {
    transaction.acceptDistributedTraceHeaders("HTTP", headers)
  }

  // Print out the distributed tracing headers
  console.log("Distributed tracing headers:")
  for (let key in headers) {
    console.info(`${key}: ${headers[key]}`)
  }
  console.log(
    "The proprietary `newrelic` header can be decoded with: `pbpaste | base64 -d | jq .`"
  )

  // Make an external HTTP request and inject distributed trace headers
  let response = await axios.get("https://example.com", { headers })

  // additional function processes
  let files = fs.readdirSync("/opt/nodejs/")
  let packageLock = fs.readFileSync("/opt/nodejs/package-lock.json").toString()
  let username = get_user()
  let password = get_pass()

  // As normal, anything you write to stdout ends up in CloudWatch
  console.info("username: ", username)
  console.info("password: ", password)
  console.info("ENVIRONMENT VARIABLES\n" + JSON.stringify(process.env, null, 2))
  console.info("EVENT\n" + JSON.stringify(event, null, 2))
  console.warn("This is a warning log.")
  console.info(files)
  console.info("***package-lock.json***\n", packageLock)
  console.info("logStream: ", context.logStreamName)

  return response.status
  // return context.logStreamName
}
